### DISCUSSION TITLE
Extracting input of sim_ber function

### DISCUSSION THREAD

**Discussion by BramvBk** [NVlabs/sionna#118] [General]
Hello,
I am using Sionna and started looking into the Toward Learned Receivers example. I would like to know if it is possible to extract and save the output data of the sim_ber function used in the evolution step of the neural receiver. so I can use the exact input for a hardware implementation to use as a comparison. 

Best regards,

**Comment by SebastianCa** [comment]
Hi,

in the [Neural Receiver](https://nvlabs.github.io/sionna/examples/Neural_Receiver.html#Evaluation-of-the-Neural-Receiver) example, the `sim_ber` function simply runs the `model` multiple times and internally evaluates the BER/BLER for a given SNR (Monte-Carlo simulation). 

If you want to store the exact neural receiver input/output, you need to modify the `E2ESystem` Model by simply returning `y` in the last line, i.e., `return b, b_hat, y`.
You can then call the new model via 
`b, b_hat, y = model(batch_size, ebno_db)`
where `b` containts the ground truth bits, `b_hat` is the neural receiver's estimate and `y` is the channel output/input to the receiver. `batch_size` and `ebno_db` depend on your desired system configuration.

**Reply by BramvBk** [reply]
Thanks for your reply i am also trying to extract the exact input data going into the neural receiver so that is  z = tf.concat([tf.math.real(y),  tf.math.imag(y),no], axis=-1) . I thought indeed sim ber just ran the model multiple times but when i try to   exports using either  tf.io.write.file or  write it to a diffrent variable (as z is reused for the diffrent resudial blocks) so i can return it in the last line it only seems to be called during the  following part:

model(1, tf.constant(10.0, tf.float32))
with open(model_weights_path, 'rb') as f:
    weights = pickle.load(f)

model.set_weights(weights)

But not when running the line 

_,bler = sim_ber(model, ebno_dbs, batch_size=64, num_target_block_errors=50, max_mc_iter=50, graph_mode=None )

**Reply by SebastianCa** [reply]
Keep in mind that the model runs in [Graph mode](https://nvlabs.github.io/sionna/examples/Sionna_tutorial_part1.html#Eager-vs-Graph-Mode) due to the `@tf.function` decorator. This means that any Python code in the `call()` function of the model will be traced only once during the first call (or if the [ input signature changes](https://www.tensorflow.org/guide/function#rules_of_tracing)).

Without knowing your exact application, I would suggest to replace the `sim_ber` function with a customized for-loop which saves the returned values & calculates the BLER manually or you could run the model in Eager mode.

**Reply by BramvBk** [reply]
thank you for your reply, I think this is indeed the issue I am having I also noticed that sim_ber has the argument graph mode that I am able to set. (graph_mode (One of [“graph”, “xla”], str) – A string describing the execution mode of mc_fun. Defaults to None. In this case, mc_fun is executed as is.) would this allow it to be traced multiple times or is this something else. can I run in eager mode by removing the decorator? (I will read into this as I am still relatively new in training nn and using tensorflow) but thanks for pointing me toward this. how would changing the sim_ber function change the execution as you explained sim_ber just runs the model multiple times so if I do the same whit a for loop how would that change . as I imagine the python code in the call is only executed the first time. or does running the customized loop cause the input signature to change?

**Reply by SebastianCa** [reply]
You need to remove the function decorator in the model. The `graph_mode` parameter in the `sim_ber` function *only* allows to run an Eager function in graph (or XLA) mode. Unfortunately, the other direction does NOT work.

If the model is properly fed with input tensors for the `batch_size` and the `snr`, the signature should not change.
You can find a detailed explanation in [Eager vs. graph mode](https://nvlabs.github.io/sionna/examples/Sionna_tutorial_part1.html#Eager-vs-Graph-Mode) for details.

**Reply by BramvBk** [reply]
thanks for your responses i will try this out
